drivers/gpio/gpio_cy8c95xx.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2021 Synopsys
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT cypress_cy8c95xx_gpio_port
+
+#include <errno.h>
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(cy8c95xx, CONFIG_GPIO_LOG_LEVEL);
+
+#include "gpio_utils.h"
+
+/** Cache of the output configuration and data of the pins. */
+struct cy8c95xx_pin_state {
+	uint8_t dir;
+	uint8_t data;
+	uint8_t pullup;
+};
+
+/** Runtime driver data */
+struct cy8c95xx_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	const struct device *i2c_master;
+	struct cy8c95xx_pin_state pin_state;
+	struct k_sem *lock;
+};
+
+/** Configuration data */
+struct cy8c95xx_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	const char *i2c_master_dev_name;
+	uint16_t i2c_slave_addr;
+	uint8_t port_num;
+};
+
+#define CY8C95XX_REG_INPUT_DATA0                0x00
+#define CY8C95XX_REG_OUTPUT_DATA0               0x08
+#define CY8C95XX_REG_PORT_SELECT                0x18
+#define CY8C95XX_REG_DIR                        0x1C
+#define CY8C95XX_REG_PULL_UP                    0x1D
+#define CY8C95XX_REG_PULL_DOWN                  0x1E
+#define CY8C95XX_REG_ID                         0x2E
+
+static int write_pin_state(const struct cy8c95xx_config *cfg,
+			   struct cy8c95xx_drv_data *drv_data,
+			   struct cy8c95xx_pin_state *pins)
+{
+	int rc;
+
+	rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				CY8C95XX_REG_PORT_SELECT, cfg->port_num);
+	if (rc == 0) {
+		i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				   CY8C95XX_REG_DIR, pins->dir);
+		i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				   CY8C95XX_REG_PULL_UP, pins->pullup);
+		i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				   CY8C95XX_REG_PULL_DOWN, ~pins->pullup);
+	}
+	rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				CY8C95XX_REG_OUTPUT_DATA0 + cfg->port_num, pins->data);
+
+	return rc;
+}
+
+static int cy8c95xx_config(const struct device *dev,
+			   gpio_pin_t pin,
+			   gpio_flags_t flags)
+{
+	const struct cy8c95xx_config *cfg = dev->config;
+	struct cy8c95xx_drv_data *drv_data = dev->data;
+	struct cy8c95xx_pin_state *pins = &drv_data->pin_state;
+	int rc = 0;
+	uint32_t io_flags;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* No strengths support */
+	if ((flags & (GPIO_DS_ALT_LOW | GPIO_DS_ALT_HIGH)) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* No open-drain support */
+	if ((flags & GPIO_SINGLE_ENDED) != 0U) {
+		return -ENOTSUP;
+	}
+
+	/* Default: pull down */
+	if ((flags & GPIO_PULL_UP) != 0) {
+		pins->pullup |= BIT(pin);
+	} else {
+		pins->pullup &= ~BIT(pin);
+	}
+
```

Comment:
Let's implement handling of `GPIO_PULL_DOWN` flag or return `-ENOTSUP` if flag is set.

Reply:
I have implemented GPIO_PULL_DOWN, if you don't set GPIO_PULL_UP, then it's GPIO_PULL_DOWN.

Reply:
Sorry, I missed that. Nevertheless, it's incorrect. Pull up, pull down have to be configured independently. Missing pull up flag does not mean that user requests pull down.

Reply:
This fragment should be written as
```
	if ((flags & GPIO_PULL_UP) != 0) {
		pins->pullup |= BIT(pin);
	} else {
		pins->pullup &= ~BIT(pin);
	}

	if ((flags & GPIO_PULL_DOWN) != 0) {
		pins->pulldown |= BIT(pin);
	} else {
		pins->pulldown &= ~BIT(pin);
	}
```
or preferably in a more concise manner using `WRITE_BIT` macro
```
	WRITE_BIT(pins->pullup, pin, (flags & GPIO_PULL_UP) != 0U);
	WRITE_BIT(pins->pulldown, pin, (flags & GPIO_PULL_DOWN) != 0U);
```
You'll need to add `pulldown` field to `struct cy8c95xx_pin_state`.